pkg/crypto/protocol/relay_difficulty.go

```diff
@@ -33,64 +33,49 @@ func IsRelayVolumeApplicable(relayHash, targetHash []byte) bool {
 // on the target number of relays we want the network to mine and the new EMA of
 // the number of relays.
 func ComputeNewDifficultyTargetHash(prevTargetHash []byte, targetNumRelays, newRelaysEma uint64) []byte {
-	// TODO_MAINNET(@red-0ne): Use a language agnostic float implementation to ensure
-	// deterministic results and avoid loss of precision. Specifically, we need to
-	// use big.Rat, delay any computation.
+	// If targetNumRelays == newRelaysEma -> do not scale -> keep the same difficulty to mine relays
+	if targetNumRelays == newRelaysEma {
+		return prevTargetHash
+	}
 
 	// Calculate the proportion of target relays relative to the EMA of actual volume applicable relays
-	difficultyScalingRatio := new(big.Float).Quo(
-		new(big.Float).SetUint64(targetNumRelays),
-		new(big.Float).SetUint64(newRelaysEma),
-	)
 	// If difficultyScalingRatio < 1 -> scale down -> increase difficulty to mine relays
 	// If difficultyScalingRatio > 1 -> scale up -> decrease difficulty to mine relays
-	if difficultyScalingRatio.Cmp(big.NewFloat(1)) == 0 {
-		return prevTargetHash
-	}
-
-	// You can't scale the base relay difficulty hash below BaseRelayDifficultyHashBz
-	isDecreasingDifficulty := difficultyScalingRatio.Cmp(big.NewFloat(1)) == 1
-	if isDecreasingDifficulty && bytes.Equal(prevTargetHash, BaseRelayDifficultyHashBz) {
+	difficultyScalingRatio := big.NewRat(int64(targetNumRelays), int64(newRelaysEma))
+	scaledDifficultyHashBz := ScaleRelayDifficultyHash(prevTargetHash, difficultyScalingRatio)
+	// If scaledDifficultyHash is longer than BaseRelayDifficultyHashBz, then use
```

Comment:
Are you sure about this?

Let's assume BaseRelayDifficulty is `fff` (allow all relays)
Let's assume ScaledDifficulty is `0001` (allow almost no relays)

We're going to get a completely wrong output.

I think my previous business logic was a bit different for this reason.
- If I'm wrong -> cool. Can you explain?
- If I'm right -> can you add a test case to mitigate the regression?

Reply:
`ScaleRelayDifficultyHash` is getting its hash bytes from a `big.Int` which should not append or pad left zero bytes.

To make it extra safe, I'll trim any extra zeros on the left side.

Reply:
This sort of things always confuses me (hard to reason about) so I'll just trust you 🤝. (I'd personally verify tests or add some if I wasn't sure)

Please let me know if you're unsure but otherwise, no blockers.